src/combinatorics/simple_graph/inc_matrix.lean

```diff
@@ -0,0 +1,409 @@
+/-
+Copyright (c) 2021 Gabriel Moise. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Gabriel Moise.
+-/
+
+import algebra.big_operators.basic
+import combinatorics.simple_graph.basic
+import data.fintype.basic
+import data.sym2
+import linear_algebra.matrix
+
+/-!
+# Incidence matrices
+
+This module defines the incidence matrix `inc_matrix` of an undirected graph `simple_graph`, and provides
+theorems and lemmas connecting graph properties to computational properties of the matrix. It also
+defines the notion of `orientation` for a `simple_graph`, picking a direction for each undirected
+edge in the graph and then defining the oriented incidence matrix `oriented_inc_matrix` based on that.
+
+## Main definitions
+
+* `inc_matrix` is the incidence matrix `M` of a `simple_graph` with coefficients in a given ring R.
+* `orientation` is a structure that defines a choice of direction on the edges of a `simple_graph`.
+* `oriented_inc_matrix` is the oriented incidence matrix `N(o)` of a `simple_graph` with
+respect to a given `orientation`.
+
+## Main statements
+
+1. ∑ e : E, M i e * M j e = 1, for any two adjacent vertices i and j.
+2. M i e * M j e = 0, for any two distinct non-adjacent vertices i, j and edge e.
+3. Every element from M is idempotent.
+4. For any vertex i, the sum on the ith row of M is equal to the degree of i.
+5. (N(o) i e) ^ 2 = M i e, for any orientation o, vertex i and edge e.
+6. For any adjacent vertices i j and edge e, N(o) i e * N(o) j e = if e = ⟦(i, j)⟧ then -1 else 0.
+7. For any non-adjacent distinct vertices i j and edge e, N(o) i e * N(o) j e = 0.
+8. (xᵀ ⬝ N(o)) e = x o.head(e) - x o.tail(e).
+
+## References
+
+<https://en.wikipedia.org/wiki/Orientation_(graph_theory)>
+-/
+
+open_locale big_operators matrix
+open finset matrix simple_graph sym2
+
+universe u
+variables {R : Type u} [ring R] [nontrivial R] [decidable_eq R]
+
+@[simp]
+lemma ite_prod_one_zero {P Q : Prop} [decidable P] [decidable Q] :
+  (ite P (1 : R) 0) * (ite Q 1 0) = ite (P ∧ Q) 1 0 :=
+by { by_cases h : P; simp [h] }
```

Comment:
This looks a bit like [`ite_mul_zero_left`](https://leanprover-community.github.io/mathlib_docs/algebra/ring/basic.html#ite_mul_zero_left), and is probably worth generalizing a little:
```
@[simp]
lemma ite_mul_ite_zero_right
  {R : Type*} [mul_zero_class R] {P Q : Prop} [decidable P] [decidable Q] (a b : R):
  (ite P a 0) * (ite Q b 0) = ite (P ∧ Q) (a * b) 0 :=
by simp [←ite_and]
```

Reply:
Note that `*` is always `mul` and never `prod`; the latter is either `×` or `∏`

Reply:
this should also not be in `simple_graph`

Reply:
Sure, but I wouldn't advise moving it until everything else looks good. Maybe stick it in a `section to_move` until the PR is otherwise ready.